### ISSUE TITLE
Redirect to the same component

### ISSUE THREAD

**Issue by DaddyWarbucks** [ReactTraining/react-router#4284]
## Version
^4.0.0-alpha.5

## Test Case
```
<Router>
    <Match pattern="/stage/:id" component={Stage} />
</Router>
```
```
class Stage extends React.Component {

  componentDidMount() {
    fetch(`/api/stage/${this.props.params.id}`).then(stage => {
      this.setState( {stage: stage} );
    });
  }

 completeStage() {
    this.setState({stageIsComplete: true});
  }

  render() {
    // If this stage is complete, redirect to the next stage, which uses the same component
    if (this.state.stageIsComplete) {
      return ( <Redirect to="/stage/someOtherId"> )
    } else {
      return ( <button onClick={this.completeStage.bind(this)}>Complete Stage</button> )
    }
 }
}
```

## Steps to reproduce

The user navigates to `/stage/123`, then clicks on "Complete Stage"

## Expected Behavior

The `Stage` component should fully "remount" itself, fetch the proper stage info, render the new stage info.

## Actual Behavior

The `Stage` component "re-renders" with the current state, rather than "remounts" itself and runs through the entire lifecycle. It does receive new props, however.
In Short, when `<Redirect>`'ing to a pattern that uses the same component as the currently rendered component, RRV4 does not remount. 
I can confirm that `<Link>` does work as expected in this same context. Replacing the `<Redirect>` with a `<Link>` (same pattern) does work (after the user clicks on it)

**Comment by nygardk** [comment]
I just ran to this exact same problem.

I'm creating a HOC that allows the wrapped component to call `setQueryParam` to set a query param to the route and `getQueryParam` to read it.

```jsx
const values = arr => Object.keys(arr).map(key => arr[key]);

const queryParametrize = Wrapped => class QueryParametrized extends Component {
  static propTypes = {
    location: PropTypes.object.isRequired,
  }

  state = {}

  render() {
    const queryParams = this.props.location.query || {};

    const wrappedProps = {
      setQueryParam: (name, value) => this.setState({ [name]: value }),
      getQueryParam: name => queryParams[name],
    };

    return values(this.state).find(value => value !== queryParams[value])
      ? <Redirect to={{ pathname: '', query: this.state }} />
      : <Wrapped {...this.props} {...wrappedProps} />
  }
};
```

Usage:
```jsx
const TestComponent = ({ getQueryParam, setQueryParam }) => (
  <div>
    Parameter currently: {getQueryParam('testparam')}

    <button onClick={() => setQueryParam('testparam', '1')}>
      set to 1
    </button>

    <button onClick={() => setQueryParam('testparam', '2')}>
      set to 2
    </button>
  </div> 
);

export default queryParametrize(TestComponent);
```

**Comment by timdorr** [comment]
This is just how React works. If you update props above a component, it doesn't remount the same component type, it just provides those new props to the component. You should use the appropriate lifecycle method: `componentWillRecieveProps`.

**Comment by DaddyWarbucks** [comment]
Thanks for the response and all of the hard work on RRv4. I have really enjoyed using it thus far!

I agree that this is how core React works, and I understand you guys are shooting to make v4 more declarative like React core. But I, respectfully, disagree that this is how a router should work. Maybe it's not possible with the V4 because of how "component-like" it is.

I had actually searched through the issues and found a similar issue where you provided the same answer about `componentWillReceiveProps`. I have currently implemented that solution, and it makes for a much more complicated component lifecycle....which is understandable. 

But, using `componentWillReceiveProps` to accomplish this actually breaks the history and the user cannot use the native browser `Back` button to go back a "stage" (as per my example).

 I have also noticed that manually entering an address into the address bar with this solution has become inconsistent. I have not nailed down that exact use/test case in order to better explain it (it may be how I am using `componentWillReceiveProps`).

But, I can confirm that the user can no longer use the native `Back` button. That is a bug, in my opinion.

**Comment by kwelch** [comment]
If what you are truly after is the ability to use the back button you should update the `Redirect` to use the `push` prop. 

By default `Redirect` replace current location in history. If you want to leverage the back button you need to pass the `push` prop to `Redirect` so that the new entry is created in history.
